src/Chapter3.hs

```diff
@@ -480,6 +514,88 @@ After defining the city, implement the following functions:
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
 
+data UpToFour a
+  = One a
+  | Two a a
+  | Three a a a
+  | Four a a a a
+  deriving Show
+
+instance Foldable UpToFour where
+  foldMap f (One x           ) = f x
+  foldMap f (Two x1 x2       ) = f x2 <> foldMap f (One x1)
+  foldMap f (Three x1 x2 x3  ) = f x3 <> foldMap f (Two x1 x2)
+  foldMap f (Four x1 x2 x3 x4) = f x4 <> foldMap f (Three x1 x2 x3)
+
+newtype Castle = MkCastle {unCastle :: String}
+  deriving Show
+
+data Wall = Wall
+  deriving Show
+
+data CultureBuilding = Church | Library
+  deriving Show
+
+newtype House = House (UpToFour Person)
+  deriving Show
+
+newtype Person = Person { personName :: String }
+  deriving Show
+
+
+data SmallCity  = SmallCity { cityCastle :: Maybe Castle
+                            , cityCultureBuilding :: CultureBuilding
+                            , cityHouses :: [House]
+                            } deriving Show
+
+data BigCity = BigCity { cityCastle :: Castle
+                       , cityWall :: Maybe Wall
+                       , cityCultureBuilding :: CultureBuilding
+                       , cityHouses :: [House]
+                       } deriving Show
+
+data City
+  = NoWallCity SmallCity
+  | WallCity BigCity
```

Comment:
That's a really nice way to solve the problem :+1: Now the invalid states are indeed unrepresentable :100: 

I would simplify this solution a bit though. You have two separate records: `SmallCity` and `BigCity`. But in reality it's enough to have only one. Note, that you can have `Wall` only if you have `Castle`. So instead of having two records with duplicating fields, you can have a single city with a field like:

```haskell
cityCastleAndWalls :: Maybe (Castle, Maybe Walls)
```

of course, have a custom data type instead of a pair is better :slightly_smiling_face: And since the `Wall` type doesn't have any fields, instead of `Maybe Castle` you could just have:

```haskell
data Castle
    = NoCastle
    | Castle String
    | CastleWithWalls String
```

Just an alternative design with different tradeoffs :slightly_smiling_face:

Reply:
It's amazing how much that one small type change simplified all the code. Looks way better now :slightly_smiling_face:

I definitely like the custom type approach. I feel like a tuple at the type level is a bit of a code smell. It's just begging for its own type :laughing: